experimental/lib/Transforms/Speculation/PlacementFinder.cpp

```diff
@@ -0,0 +1,359 @@
+
+  // Mark all values that are speculative in the speculation BB
+  llvm::DenseSet<Value> specValues;
+  specValues.insert(specPos.srcOpResult);
+  markSpeculativePaths(specPos.dstOp, specValues);
+
+  // Iterate all operations in the speculation BB
+  unsigned bb = getLogicBB(specPos.dstOp).value();
+  for (Operation *blockOp : handshakeBlocks.blocks.lookup(bb)) {
+    // Create a save if an operation has both spec and non-spec operands
+    bool hasNonSpecInput = false;
+    bool hasSpecInput = false;
+    for (Value operand : blockOp->getOperands()) {
+      if (specValues.count(operand))
+        hasSpecInput = true;
+      else
+        hasNonSpecInput = true;
+    }
+
+    if (hasSpecInput and hasNonSpecInput) {
+      for (Value operand : blockOp->getOperands()) {
+        // Create a Save for every non-speculative operand
+        if (not specValues.count(operand)) {
+          // No save needed in front of Source Operations
+          if (isa<handshake::SourceOp>(operand.getDefiningOp()))
+            continue;
+
+          placements.addSave(operand, blockOp);
+        }
+      }
+    }
+  }
+
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// Commit Units Finder Methods
+//===----------------------------------------------------------------------===//
+
+void PlacementFinder::findCommitsTraversal(llvm::DenseSet<Operation *> &visited,
+                                           Operation *currOp) {
+  // End traversal if currOp is already in visited set
+  if (auto [_, isNewOp] = visited.insert(currOp); !isNewOp)
+    return;
+  for (Value res : currOp->getResults()) {
+    for (Operation *succOp : res.getUsers()) {
+      if (placements.containsSave(res, succOp)) {
+        // A Commit is needed in front of Save Operations. This will be later
+        // converted into a SaveCommit for multiple loop speculation.
+        placements.addCommit(res, succOp);
+      } else if (isa<handshake::LSQOp, handshake::MemoryControllerOp,
+                     handshake::MCStoreOp, handshake::LSQStoreOp,
+                     handshake::MCLoadOp, handshake::LSQLoadOp>(succOp)) {
+        // A commit is needed in front of memory operations
+        placements.addCommit(res, succOp);
+      } else if (isa<handshake::EndOp>(succOp)) {
+        // A commit is needed in front of the end/exit operation
+        placements.addCommit(res, succOp);
+      } else {
+        findCommitsTraversal(visited, succOp);
+      }
+    }
+  }
+}
+
+struct endpointComparator {
+  bool operator()(const BBEndpoints &a, const BBEndpoints &b) const {
+    if (a.srcBB != b.srcBB)
+      return a.srcBB < b.srcBB;
+    return a.dstBB < b.dstBB;
+  }
+};
+using BBEndpointsMap =
+    std::map<BBEndpoints, std::vector<OpPlacement>, endpointComparator>;
+
+// Data structure to hold all arcs leading to a single BB predecessor
+// Note: srcBB and dstBB can be equal when the arcs are Backedges
+struct BlockPredecessor {
```

Comment:
naming here confuses me "a single BB predecessor", then both src and dst are stored. 
what does "leading to a predecessor" mean? is it from a predecessor? if it has both a src and a dest does it hold "all" the arcs from the src or just the ones to the dest?
This is a group of arcs that have the same source and dest? I don't see how predecessors are involved

Reply:
I think this struct should be called a bbArc, and the vector is instead edges.

Reply:
The naming you suggest is very appropriate. The whole section has been refactored with your suggestions.